### ISSUE TITLE
java.lang.RuntimeException: No OpenGL context found in the current thread.

### ISSUE THREAD

**Issue by ghost** [nifty-gui/nifty-gui#111]
When I try and give a nifty panel (named "myPanel") keyboard focus, I'm getting a strange exception that doesn't make any sense because I'm not doing any multithreading, and the OpenGL context is active in my (only) thread. The panel "myPanel" exists in myScreen.xml. I only try and set the keyboard focus when myScreen is active.

What I'm trying to achieve is to take the keyboard focus away from a nifty-textfield (i.e., blinking cursor) when myPanel is left clicked on, and give it to myPanel. Just setting myPanel to focusable="true" in myPanel.xml isn't good enough (although I do have this set) because I want the keyboard focus to happen when the mouse left clicks on myPanel.

Ok, so what I'm REALLY trying to achieve is taking the keyboard focus away from nifty components such as textfield and giving it to the game. I am just using lwjgl and nifty, not slick or jme. It doesn't work automatically because nifty only releases textfield keyboard focus when clicking on ANOTHER nifty component. So I have the 100% transparent panel named "myPanel" overlay the game to take keyboard focus. Otherwise if you type in a nifty textfield and then click inside the game with the mouse, you cannot use the keyboard inside the game; it will still type in the textfield.

Setting visibleToMouse="true" is also not an option because I have openGL rendered graphics underneath myPanel that need to be interacted with the mouse. I know you changed the behavior of visibleToMouse to fix some bugs, but this change is really hurting me because if it worked like before, then I could set visibleToMouse="true" on myPanel and get the behavior I want for free, and still get mouse clicks underneath the panel. Is there a way to force the mouse clicks to not be consumed only for that one panel?

Anyway, everything else is working correctly (nifty, lwjgl, openGL). If I remove this line from MyScreenController.java:

```
this.myScreen.findElementByName("myPanel").setFocus();
```

The problem goes away. Here is the stack trace that the above line causes:

<pre>
java.lang.RuntimeException: No OpenGL context found in the current thread.
java.lang.RuntimeException: No OpenGL context found in the current thread.
    at org.lwjgl.opengl.GLContext.getCapabilities(GLContext.java:124)
    at org.lwjgl.opengl.GL11.glDeleteTextures(GL11.java:731)
    at de.lessvoid.nifty.renderer.lwjgl.render.LwjglRenderImage.dispose(LwjglRenderImage.java:65)
    at de.lessvoid.nifty.render.NiftyImageManager$ReferencedCountedImage.removeReference(NiftyImageManager.java:107)
    at de.lessvoid.nifty.render.NiftyImageManager.dispose(NiftyImageManager.java:52)
    at de.lessvoid.nifty.render.NiftyRenderEngineImpl.disposeImage(NiftyRenderEngineImpl.java:559)
    at de.lessvoid.nifty.render.NiftyImage.dispose(NiftyImage.java:84)
    at de.lessvoid.nifty.effects.impl.ImageOverlayPulsate.deactivate(ImageOverlayPulsate.java:50)
    at de.lessvoid.nifty.effects.Effect.setActiveInternal(Effect.java:139)
    at de.lessvoid.nifty.effects.Effect.deactivate(Effect.java:132)
    at de.lessvoid.nifty.effects.EffectProcessorImpl.reset(EffectProcessorImpl.java:107)
    at de.lessvoid.nifty.effects.EffectProcessorImpl.setActive(EffectProcessorImpl.java:179)
    at de.lessvoid.nifty.effects.EffectManager.stopEffect(EffectManager.java:125)
    at de.lessvoid.nifty.elements.Element.stopEffectInternal(Element.java:1259)
    at de.lessvoid.nifty.elements.Element.stopEffect(Element.java:1244)
    at de.lessvoid.nifty.controls.textfield.TextFieldControl.onFocus(TextFieldControl.java:308)
    at de.lessvoid.nifty.controls.NiftyInputControl.onFocus(NiftyInputControl.java:88)
    at de.lessvoid.nifty.elements.Element.stopEffectInternal(Element.java:1271)
    at de.lessvoid.nifty.elements.Element.stopEffect(Element.java:1244)
    at de.lessvoid.nifty.controls.FocusHandler.setKeyFocus(FocusHandler.java:215)
    at de.lessvoid.nifty.elements.Element.setFocus(Element.java:1883)
</pre>

**Comment by ghost** [comment]
I'm also now getting this same exception now when calling Nifty's TextCreator.create method.

Caused by: java.lang.RuntimeException: java.lang.RuntimeException: No OpenGL context found in the current thread.
    at de.lessvoid.nifty.Nifty.addControlsWithoutStartScreen(Nifty.java:425)
    at de.lessvoid.nifty.controls.dynamic.TextCreator.create(TextCreator.java:32)
Caused by: java.lang.RuntimeException: No OpenGL context found in the current thread.
    at org.lwjgl.opengl.GLContext.getCapabilities(GLContext.java:124)
    at org.lwjgl.opengl.GL11.glGenLists(GL11.java:1390)
    at de.lessvoid.nifty.renderer.lwjgl.render.font.Font.initDisplayList(Font.java:91)
    at de.lessvoid.nifty.renderer.lwjgl.render.font.Font.init(Font.java:74)
    at de.lessvoid.nifty.renderer.lwjgl.render.LwjglRenderFont.<init>(LwjglRenderFont.java:14)
    at de.lessvoid.nifty.renderer.lwjgl.render.LwjglRenderDevice.createFont(LwjglRenderDevice.java:198)
    at de.lessvoid.nifty.render.ScalingRenderDevice.createFont(ScalingRenderDevice.java:32)
    at de.lessvoid.nifty.render.NiftyRenderEngineImpl.createFont(NiftyRenderEngineImpl.java:192)
    at de.lessvoid.nifty.loaderv2.types.apply.Convert.font(Convert.java:44)
    at de.lessvoid.nifty.loaderv2.types.apply.ApplyRenderText.apply(ApplyRenderText.java:23)
    at de.lessvoid.nifty.elements.Element.initializeFromAttributes(Element.java:433)
    at de.lessvoid.nifty.loaderv2.types.ElementType.applyAttributes(ElementType.java:218)
    at de.lessvoid.nifty.loaderv2.types.ElementType.applyStandard(ElementType.java:172)
    at de.lessvoid.nifty.loaderv2.types.ElementType.create(ElementType.java:144)
    at de.lessvoid.nifty.controls.dynamic.attributes.ControlAttributes.buildControl(ControlAttributes.java:391)
    at de.lessvoid.nifty.controls.dynamic.attributes.ControlAttributes.createText(ControlAttributes.java:344)
    at de.lessvoid.nifty.controls.dynamic.TextCreator.access$000(TextCreator.java:12)
    at de.lessvoid.nifty.controls.dynamic.TextCreator$1.createControl(TextCreator.java:29)
    at de.lessvoid.nifty.Nifty$ControlToAdd.createControl(Nifty.java:997)
    at de.lessvoid.nifty.Nifty.addControlsWithoutStartScreen(Nifty.java:423)

**Comment by ghost** [comment]
Ok, now that the exception happened in two different places, I think I figured out the commonality between them: both of the exceptions originate (in different places) in code that is inside a custom EventBus subscriber method, i.e., an event hander. I guess I could theoretically imagine events being fired off before the openGL context is created. But the thing is, that can't be what's causing this because these exceptions are being thrown DURING the middle of a game, and the OpenGL context is created before any game is ever started.

But I definitely think I'm on to something regarding the EventBus. I just don't know what exactly.

**Comment by ghost** [comment]
After an hour of research and testing I've got some more understanding on this issue. I caught the above exception and checked the values of Display.isCreated and Display.isCurrent. When the above exception is thrown, Display.isCreated is true but Display.isCurrent is false. How can I make niftyGui render on the correct (main) thread (that contains the opengl context) from an EventBus subscriber / callback method?

**Comment by ghost** [comment]
Ok, so I guess once again this is not really a niftyGui problem. In the EventBus subscriber, I am going to create a Runnable with the niftyGui code in it, queue it, and then run it from the main loop, like maybe right before nifty.render() or something.

It still seems like this could be a nifty problem though because the EventBus I'm using is the same EventBus used in nifty. So this same problem could potentially happen to anyone that subscribes to a nifty event and then tries to modify something like an image that requires an OpenGL call via the underlying renderer implementation. Or does nifty customize EventBus in some way to account for this issue? I'd like to know because maybe I could use the same solution, and maybe it would be more elegant than mine...

**Comment by void256** [comment]
Sorry for the late response. I was busy with other non-Nifty related stuff, unfortunately. But I'm back now!

Do you use threads in your application?

In general the current Nifty version is not aware of a multithreaded environment and assumes there is only a single thread.

For the EventBus, Nifty creates its own individual EventService. This is how this special "NiftyEventBus" is created when a new Nifty instance is created:

``` java
EventServiceLocator.setEventService("NiftyEventBus", new ThreadSafeEventService());
```

According to http://eventbus.org/usage.html for Non-Swing Applications this works like so:

> The ThreadSafeEventService "extends the calling stack" when subscribers are called, i.e. subscribers are called on the same thread that the event is published on.

So in general a publish() of an Event on Thread A should call all of the subscribers on the same Thread A. If you publish from a Thread B then the events would be received on Thread B as well. If Thread B is not the same as the Thread that has created the Nifty instance I think you'd see exactly the effects you've experienced.

**Comment by ghost** [comment]
Thanks for the reply. I am trying to use the default EventBus service for my own purposes IN ADDITION to the "NiftyEventBus" service.

Is this possible? For example, setting:

```
EventServiceLocator.setEventService (EventServiceLocator.SERVICE_NAME_SWING_EVENT_SERVICE, new ThreadSafeEventService());
```

I thought adding the above line of code would make this exception go away but it doesn't. Not sure if it's conflicting somehow with the "NiftyEventBus"?

And no, as I mentioned above I'm not doing any (explicit) multithreading,

**Comment by ghost** [comment]
I think I'm going to answer my own question here. Yes, I can run multiple EventService implementations at the same time because the subscriber annotations are different for each one. NiftyEventBus uses a custom annotation, @NiftyEventSubscriber, the Swing event service uses the @EventSubscriber annotation.

Furthermore, setting:

```
EventServiceLocator.setEventService (EventServiceLocator.SERVICE_NAME_SWING_EVENT_SERVICE, new ThreadSafeEventService());
```

is working perfectly. While I am still getting the same exception as before, after adding the above setting, while the nifty portion of the stack trace remained the same, the non-nifty part changed slightly, and now includes the kryonet library, but no longer includes anything regarding the EventDispatchThread as before. Bam!

The above call is causing the Swing event service to call the subscribers on the same thread as the publishing method. But now the problem has flip-flopped because the EventBus is doing exactly what I asked it to, but this is causing a problem with the kryonet library's event listeners! Kryonet starts up separate client / server threads, from which it calls its event listeners. These event listeners are calling other methods, which handle the events by publishing to the swing event service implementation of the EventBus. So since the publishing is not happening on my main openGL thread, the subscribers are also not getting called from my main openGL thread, which is why nifty is still throwing an exception.

So I've narrowed this down now to a non-nifty issue. Thanks for your help @void256. This has been a serious pain to figure out, and your comment has really helped me. I still am not sure how to tell the EventBus to publish to the main thread in the case of kryonet, without having to resort to queuing Runnable jobs to be executed later by the main thread. If you know a better way (i.e. EventBus configuration) I'd be very grateful, but you can close this now since it's not a nifty issue - once again ;-).

Note: To avoid any confusion, when I say "swing event service" above, I mean a default ThreadSafeEventService implementation (as opposed to the NiftyEventBus ThreadSafeEventService). My program doesn't use Swing at all.

**Comment by void256** [comment]
Nice!

Well, I'd suggest you look at the source code of the EventBus class org.bushe.swing.event.SwingEventService. This class simply extends the ThreadSafeEventService and makes sure that the publish() method executes events on the EventDispatchThread (EDT) of Swing.

I'm pretty sure you can apply a similar solution to your problem. Just make a new subclass of ThreadSafeEventService and schedule new Runnable jobs at whatever Thread you want :)

Good luck! I'll close this Issue now but please feel free to comment what you used at the end.
